BAU: First steps for local startup & speed up build pipeline

CharlesIC opened the pull request:
This is initial work for the upcoming local startup with Docker.
Mainly some tidying up and optimising of the current builds. Recommended to review by PRs.

**Summary of changes** (see individual commit messages for details):
  - Move everything local start-up related to one directory
  - Add missing ASCII art banners
  - Use console logging locally and in tests
  - Restructure docker images to make use of composition and caching
  - Adjust the build pipeline and optimise the build by using caching
  - Don't log ingress/egress messages on Stub Connector resources
 - Use generated values for connector and gateway endpoints in tests
 - Run unit tests by default when building the apps

CharlesIC committed:
BAU: Move everything local start-up related to one directory

Step 1: Move everything local startup related to its own directory.
These scripts are not currently working but will be kept for reference for the time being until we have a working solution.
We may be able to reuse some of the existing stuff so it's worth keeping it temporarily.

Author: @CharlesIC

CharlesIC committed:
BAU: Add missing ASCII art banners

Step 2: Add ✨sparkle✨ to app logs
This is a Dropwizard tradition, so let's follow it 🙌

Author: @CharlesIC

CharlesIC committed:
BAU: Make builds faster and the build pipeline more logical and efficient

1. Revamp the Docker images for our apps
We now have two base images for two types of services: regular Java apps (root Dockerfile) and Java apps that need the HSM client libraries (cloudhsm/base-images/jdk11-jre/Dockerfile). Apps that need to talk to the HSM also need the HSM JCE libraries at the build stage and there's now a new base image for that (cloudhsm/base-images/jdk11-jce-gradle/Dockerfile).

Also made a number of small improvements to the Docker images to make builds faster.

2. Optimise the pipeline

  - Reflect changes in the structure of the Docker images
  - Use caching when build docker images
  - Don't run `docker pull` after building or `get`ting a docker image resource - those caches is not used for anything so we can save some time by not pulling them
  - Restructure the pipeline to build the base images such as the HSM and VSP separately as they will rarely change and also don't logically belong to the Proxy Node build
  - Add job groups
  - Don't re-set default values in the integration pipeline

Author: @CharlesIC

CharlesIC committed:
BAU: Use console logging in tests and local startup by default

This makes the logs developer-readable when running the tests or running the apps locally

Author: @CharlesIC

philandstuff commented:
Looks great from a quick on-📱eyeball review. I’ll let more thorough reviewers add ✅ though

JohnWattsGDS reviewed (dismissed):
So, we're still running as root, just implicitly?

CharlesIC commented:
Yep - that's the image default. 
We should manually create a non-privileged user and use that instead, as per the [Docker best practices guide](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user)

tomrosier commented on cloudhsm/base-images/jdk11-jce-gradle/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+ARG JAVA_HOME=/opt/java/openjdk
+ARG GRADLE_HOME=/opt/gradle
+FROM gradle:jdk11 AS build
```
This is redundant it will get overridden by line 7

tomrosier commented on cloudhsm/base-images/jdk11-jce-gradle/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+ENV JAVA_HOME $JAVA_HOME
+ENV GRADLE_HOME $GRADLE_HOME
+
+COPY --from=build $JAVA_HOME $JAVA_HOME
```
Could do with a comment to say this is being taken from the `FROM gradle:jdk11 AS build`

CharlesIC replied:
It says `--from=build` and `AS build` at the top of the file - this is standard Dockerfile syntax. Do we really need a comment for that? I think it's clear from the context

CharlesIC committed:
BAU: Don't set user as root explicitly

This is the default and not a good practice anyway. In fact, we should run the apps with a non-privileged user.

Author: @CharlesIC

CharlesIC committed:
BAU: Don't log ingress/egress messages on Stub Connector resources

This is not a real service we have to monitor

Author: @CharlesIC

CharlesIC committed:
BAU: Use generated values for connector and gateway endpoints in tests

These values are generated in the templates/helpers file and should be used to run tests.

Author: @CharlesIC

CharlesIC committed:
BAU: Run tests by default when building the apps

Author: @CharlesIC

JohnWattsGDS reviewed (approved):
```
    __    ______________  ___
   / /   / ____/_  __/  |/  /
  / /   / / __  / / / /|_/ / 
 / /___/ /_/ / / / / /  / /  
/_____/\____/ /_/ /_/  /_/   
                             
```

Wynndow commented on cloudhsm/base-images/jdk11-jce-gradle/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+FROM amazonlinux:2.0.20190508
+
+# Install Gradle and Java
+ARG JAVA_HOME
```
Might be my lack of understanding of Dockerfiles but do you need lines 10 and 11 here considering these have already been defined on lines 3 and 4?

CharlesIC replied:
Yep - you need to redefine the `ARG`s at every build stage as they're not propagated
